File: src/ua/inf/krre/aprilbrush/view/PaintView.java
package ua.inf.krre.aprilbrush.view;

import ua.inf.krre.aprilbrush.logic.BrushEngine;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class PaintView extends View {
    private BrushEngine engine;
    private Bitmap bitmap;
    private Canvas canvas;
    private Paint bitmapPaint;

    public PaintView(Context context, AttributeSet attrs) {
        super(context, attrs);
        bitmapPaint = new Paint(Paint.DITHER_FLAG);
        engine = new BrushEngine();
    }

    @Override
    public void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
        canvas = new Canvas(bitmap);
    }

    @Override
    public void onDraw(Canvas canvas) {
        canvas.drawColor(Color.LTGRAY);
        canvas.drawBitmap(bitmap, 0, 0, bitmapPaint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        engine.paintDab(canvas, event);
        invalidate();
        return true;
    }

}

File: AprilBrush/src/main/kotlin/view/PaintView.kt
package org.krre.aprilbrush.view

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.util.Log
import org.krre.aprilbrush.logic.BrushEngine
import org.krre.aprilbrush.drawable.Dab
import android.graphics.Bitmap

class PaintView(context : Context, attrs : AttributeSet) : View(context, attrs) {
    private val TAG = "AB"
    private val bufferPaint : Paint = Paint()
    private val brushEngine = BrushEngine()
    private val dabDrawable = Dab();
    {
        brushEngine.dab = dabDrawable
    }

    override fun onSizeChanged(w : Int, h : Int, oldw : Int, oldh : Int) {
        super.onSizeChanged(w, h, oldw, oldh);
        brushEngine.setBufferSize(w, h)
    }

    override fun onDraw(canvas : Canvas ) {
        canvas.drawColor(Color.WHITE);
        canvas.drawBitmap(brushEngine.bufferBitmap!!, 0f, 0f, bufferPaint)
    }

    override fun onTouchEvent(event : MotionEvent ) : Boolean {
        when (event.getAction()) {
            MotionEvent.ACTION_DOWN -> {
                Log.d(TAG, "down")
                brushEngine.paintDab(event)
                invalidate()
            }
            MotionEvent.ACTION_MOVE -> {
                brushEngine.paintDab(event)
                invalidate()
            }
            MotionEvent.ACTION_UP -> {
                Log.d(TAG, "up")
                invalidate()
            }
        }
        return true;
    }
}
